Finish revival mechanic

kenneth-fung opened the pull request:
Implemented
- Interaction time
  - Each interaction can now be set to take a certain amount of time (currently, mounting takes 0 seconds and revival takes 3 seconds)
  - The interaction button `E` must be held for the set time before the interaction will complete
  - If an interaction takes more than 0 seconds, the player will be forced to stop moving while interacting
- Interaction UI prompt
  - When near an interaction, the player will see a UI prompt appear that says the key to press and the interaction that will result
  - If multiple interactions are within range, the prompt will only show for the closest one
  - If an interaction takes more than 0 seconds, a progress bar will appear beneath the key to press; the bar will fill while the interaction key is held down
- Action state machine
  - The `CombatStateMachine` has been renamed to `ActionStateMachine` and `CombatState` has been renamed to `ActionState`, as the new `InteractState` uses that state machine and calling it a combat state just seems wrong

Bug Fixes
- Fix dragon regenerating mana while dead
- Fix dragon animation falling out of sync if it dies while touching the ground
- Fix ranged dragon enemy missing walk/run animations

Issues
- `ActionState` still takes a `Combat` component as its owner when it should probably take an `ActorController`
  - This would require a serious amount of work for little pay-off, meaning it will not be fixed for this project
- `ActionStateMachine` is stored in the `Combat` component when it should likewise probably be stored in the `ActorController` component
  - This will not be done for this project

kenneth-fung committed:
Move movement states into subfolder

kenneth-fung committed:
Add duration for interactions

This commit makes each interaction take a specified amount of time to
complete. Mounting takes 0 seconds and reviving takes 3 seconds. The
player must hold down the interact button for the specified time before
the interaction completes.

A new InteractState is created and used in the combat state machine
during interaction, which is not ideal. Combat states/combat state
machine might be renamed in a future commit.

Let's
- create an InteractState as a child of CombatState
- make each Interactable take a certain amount of time
- allow interactions to be cancelled mid-interaction
- add necessary inputs to start, stop, and cancel interactions

kenneth-fung committed:
Interrupt interaction upon hurt or death

kenneth-fung committed:
Prevent actors from interacting simultaneously

This commit adds a check to make sure the same interactable cannot be
interacted with simultaneously by multiple actors.

kenneth-fung committed:
Add death animtion end event

This commit prepares the way for not enabling the revive interactable
until the actor's death animation has completed. Prefabs will need to
be modified to do that (subscribing to event in the inspector).

kenneth-fung committed:
Create draft knight/dragon prefabs with updates

The changes made to draft knight/dragon will be applied to the original
prefabs during merge.

kenneth-fung committed:
Implement partial interactable prompt UI

The prompt UI is incorrectly placed. Actor prefabs will need to be
restructured for placement to be correct.

kenneth-fung committed:
Merge branch 'master' into revival

kenneth-fung committed:
Restructure actor prefabs

Actor prefabs have been restructured so that UI objects can be attached
to actors without having them flipped on the X axis when the actor turns
left. This commit also deletes the Draft Knight/Dragon prefabs, applying
their changes to existing knight and dragon prefabs.

Let's
- add an Actor child object to all actor prefabs
- make objects holding components that benefit from scale flipping
  children of the Actor child object
- in Movement component, flip the Actor child object instead of the
  Movement component's transform
- use RPC flagged in child Movement classes to sync flipping of local
  scale

kenneth-fung committed:
Flip interaction prompt about Y axis

The interaction prompts are now flipped across the Y axis when the actor
changes direction.

Let's
- add an event to the Movement component that is invoked when an actor
  flips direction
- add a method, FlipPrompt, to the Interactable UI component that flips
  the prompt about the Y axis
- subscribe the method to the event

kenneth-fung committed:
Fix dragon regenerating mana after dying

Let's
- make toggling of combat abilities direct method calls instead of event
  listeners
- remove invocation of dismount event from MountDeathHandler method in
  mount interactable

kenneth-fung committed:
Fix dragon animation unsyncing on death

Previously, the dragon could fail to start its death animation upon
dying if it was touching the ground when entering the death state. This
was caused by a lack of animator transition directly to the death state
if it was not falling.

kenneth-fung committed:
Accept interactables in detector if local detector

This commit makes interactable detectors ignore interactables if the
detector is not owned by the local client.

kenneth-fung committed:
Disable interactable detector when actor is dead

The current nearest interactable prompt is hidden upon death.

kenneth-fung committed:
Update prompts when interactable status updates

This commit adds logic to update the nearest interactable (and thus the
displayed interactable prompts) when an interactable in range updates
its enabled/disabled status.

kenneth-fung committed:
Stop movement if interaction takes time

kenneth-fung committed:
Hide mount prompt when mounted

Mount interactable is actually disabled while mounted. The interact
button now doubles as the dismount button for the knight.

kenneth-fung committed:
Display progress bar while interacting

A progress bar will now appear beneath the interaction key prompt if the
interaction takes more than 0 seconds.

kenneth-fung committed:
Reset progress bar fill if interaction interrupted

kenneth-fung committed:
Rename combat state to action state

The combat state machine has also been renamed to action state machine.

kenneth-fung committed:
Merge branch 'master' into revival

kenneth-fung committed:
Fix interaction bar resetting to zero on end

The interaction bar would visibly reset to 0 for a split second at the
end of an interaction before the prompt disappeared. This commit fixes
that.

kenneth-fung committed:
Remove unnecessary parameters in node method

This commit removes some unnecessary parameters in two private methods
in the SetRangedAttackPosNode.

kenneth-fung committed:
Fix missing ranged dragon enemy anim transition

The ranged dragon enemy was missing a transition in its animator
controller from the idle state to movement state, causing it to slide
across the ground instead of walking/running. This has been fixed.

kenneth-fung committed:
Increase ranged dragon enemy max range

This helps compensate for the huge height at which the dragon can ascend
to.

Eriksen2411 reviewed (approved):
LGTM

lowkangn reviewed (changes requested):
There is an issue with the Knight not facing the same direction as the Dragon when mounted. The position of the Knight's sprite is also offset when mounting the Dragon while it is facing left.
Otherwise, the rest of it appears to work fine.

kenneth-fung committed:
Fix mounted knight offset and facing direction

Let's
- add a handler in mount interactable for responding to a flip in
  direction
- subscribe the handler to the dragon's flip direction event in the
  inspector
- expose an actor's photon view ID from movement component
- flip the mounted rider upon the mount flipping in mount interactable
  component, including manually setting of local position offset
- disable rider's photon view transform syncing while mounted in ground
  movement component's mount/dismount RPC calls

lowkangn reviewed (approved):
LGTM
